File: LuceneAdventureWorks.Common/Search/DirectoryFactory.cs
using Lucene.Net.Store;
using Directory = Lucene.Net.Store.Directory;

namespace LuceneAdventureWorks.Common.Search;

public static class DirectoryFactory
{
    public static Directory Create(string path)
    {
        var directoryType = Environment.GetEnvironmentVariable("LUCENE_DIRECTORY_TYPE");

        var directory = directoryType?.ToLowerInvariant() switch
        {
            // "ram" => new RAMDirectory(), // TODO: support this, but it doesn't make sense with current code
            "simplefs" => new SimpleFSDirectory(new DirectoryInfo(path)),
            "niofs" => new NIOFSDirectory(new DirectoryInfo(path)),
            "mmap" => new MMapDirectory(new DirectoryInfo(path)),
            null => FSDirectory.Open(new DirectoryInfo(path)), // use default for OS
            _ => throw new InvalidOperationException($"Unknown directory type: {directoryType}")
        };

        Console.WriteLine($"Created {directory.GetType().Name} for {path}");

        return directory;
    }
}

File: java/LuceneAdventureWorks/src/main/java/org/example/DirectoryFactory.java
package org.example;

import org.apache.lucene.store.*;

import java.io.File;
import java.io.IOException;
import java.util.Locale;

public final class DirectoryFactory {
    public static Directory Create(String path) throws IOException {
        var directoryType = System.getenv("LUCENE_DIRECTORY_TYPE");

        Directory directory;
        if (directoryType == null) {
            directory = FSDirectory.open(new File(path));
        } else {
            switch (directoryType.toLowerCase(Locale.ROOT)) {
                case "simplefs":
                    directory = new SimpleFSDirectory(new File(path));
                    break;
                case "niofs":
                    directory = new NIOFSDirectory(new File(path));
                    break;
                case "mmap":
                    directory = new MMapDirectory(new File(path));
                    break;
                default:
                    throw new java.lang.IllegalArgumentException("Unknown directory type: " + directoryType);
            }
        }

        System.out.println("Created " + directory.getClass().getName() + " for " + path);

        return directory;
    }
}
